analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Invocation.cs

```diff
@@ -1082,10 +1082,110 @@ public void Main()
-            validator.ValidateTag("InstanceTwo", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
-            validator.ValidateTag("NoArgs", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
-            validator.ValidateTag("MoreArgs", x => x.AllConstraints.Should().ContainSingle().Which.Should().Be(ObjectConstraint.NotNull));
+            validator.ValidateTag("InstanceOne", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+            validator.ValidateTag("InstanceTwo", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+            validator.ValidateTag("NoArgs", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+            validator.ValidateTag("MoreArgs", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull));
+        }
+
+        [DataTestMethod]
+        [DataRow("null", "null", true, ConstraintKind.ObjectNull, ConstraintKind.ObjectNull)]
+        [DataRow("null", "new object()", false, ConstraintKind.ObjectNull, ConstraintKind.ObjectNotNull)]
+        [DataRow("new object()", "null", false, ConstraintKind.ObjectNotNull, ConstraintKind.ObjectNull)]
+        [DataRow("new int?()", "null", true, ConstraintKind.ObjectNull, ConstraintKind.ObjectNull)]
+        [DataRow("(int?)null", "null", true, ConstraintKind.ObjectNull, ConstraintKind.ObjectNull)]
+        [DataRow("new int?(42)", "null", false, ConstraintKind.ObjectNotNull, ConstraintKind.ObjectNull)]
+        [DataRow("new int?()", "42", false, ConstraintKind.ObjectNull, ConstraintKind.ObjectNotNull)]
+        public void Invocation_ReferenceEquals_LearnResult(string left, string right, bool expectedResult, ConstraintKind expectedConstraintLeft, ConstraintKind expectedConstraintRight)
+        {
+            var code = $"""
+                object left = {left};
+                object right = {right};
+                var result = ReferenceEquals(left, right);
+                Tag("Result", result);
+                Tag("Left", left);
+                Tag("Right", right);
+                """;
+            var validator = SETestContext.CreateCS(code).Validator;
+            validator.ValidateTag("Result", x => x.HasConstraint(BoolConstraint.From(expectedResult)).Should().BeTrue());
+            validator.ValidateTag("Left", x => x.AllConstraints.Select(x => x.Kind).Should().ContainSingle().Which.Should().Be(expectedConstraintLeft));
+            validator.ValidateTag("Right", x => x.AllConstraints.Select(x => x.Kind).Should().ContainSingle().Which.Should().Be(expectedConstraintRight));
+        }
+
+        [DataTestMethod]
+        [DataRow("new object()", "new object()")]
+        [DataRow("new object()", "Unknown<object>()")]
+        [DataRow("Unknown<object>()", "new object()")]
+        [DataRow("Unknown<object>()", "Unknown<object>()")]
+        [DataRow("new int?(42)", "Unknown<int?>()")]
+        [DataRow("(int?)42", "(int?)42")]
```

Comment:
And here you can't learn between two boxed `int` because you can never know if they point to the same heap location or not. 
More in general, you can't learn on value types passed to `ReferenceEquals` (the "anomalous" behavior reported [here](https://learn.microsoft.com/en-us/dotnet/api/system.object.referenceequals?view=net-7.0#remarks)).
Is that correct?

If so, you should also have the same behavior for non-nullable structs, such as `int`, right? For example: `[DataRow("42", "42")]`.

Reply:
We don't learn, because we don't even try. We only care about comparing to `Null` at this point.

We will eventually add support for some kind of NumberConstraint in the CBDE replacement, and on that one, we will be able to properly learn on `Equals(42, 42)`. And we will have to change the implementation for `RerefernceEquals` to avoid learning, because of the reason you've mentioned.